Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/ServerHandler.java

```diff
@@ -0,0 +1,124 @@
+package com.example.android.camera2.slowmo;
+
+import android.content.ContentValues;
+import android.os.AsyncTask;
+import android.util.Log;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.PrintWriter;
+import java.net.Socket;
+import java.util.ArrayList;
+import java.util.concurrent.ExecutionException;
+import org.jetbrains.annotations.NotNull;
+
+/** Handles every interaction of app with the server. */
+public class ServerHandler {
+
+  private Integer serverSocketPort;
+  private long serverStartTimestamp;
+  private Socket socket;
+  private BufferedReader inputReader;
+  private PrintWriter outputWriter;
+  private long hostSyncTimestamp;
+
+  public ServerHandler(Integer serverSocketPort) {
+    this.serverSocketPort = serverSocketPort;
+  }
+
+  public long getHostSyncTimestamp() {
+    return hostSyncTimestamp;
+  }
+
+  public Integer getServerSocketPort() {
+    return serverSocketPort;
+  }
+
+  public void setServerSocketPort(Integer serverSocketPort) {
+    this.serverSocketPort = serverSocketPort;
+  }
+
+  public long getServerStartTimestamp() {
+    return serverStartTimestamp;
+  }
+
+  public void startConnection() {
+    ConnectionThread connectionThread = new ConnectionThread();
+    connectionThread.execute();
+  }
+
+  public void sendKeySimulationSignal() {
+    hostSyncTimestamp = System.currentTimeMillis();
+    outputWriter.write("started capture*");
+    outputWriter.flush();
+  }
+
+  @NotNull
+  public long getSyncOffset() {
+    return serverStartTimestamp - hostSyncTimestamp;
+  }
+
+  /** Creates a socket for communication with the server. Sets reader and writer for server
+   * connection. */
+  private class ConnectionThread extends AsyncTask<Void, Void, Void> {
+
+    @Override
+    protected Void doInBackground(Void... voids) {
+      try {
+        String SERVER_IP = "127.0.0.1";
+        socket = new Socket(SERVER_IP, serverSocketPort);
+        outputWriter = new PrintWriter(socket.getOutputStream());
+        inputReader = new BufferedReader(new InputStreamReader(socket.getInputStream()));
+        Log.d(ContentValues.TAG, "Connection established");
+      } catch (IOException e) {
+        e.printStackTrace();
+      }
+      return null;
+    }
+  }
+
+  public ArrayList<Long> downloadServerTimeStamps() {
+    if (serverSocketPort == null || !socket.isConnected()) {
+      Log.d(ContentValues.TAG, "Can't find server to download timestamps. No port bound or connection is dead.");
+      return null;
+    }
+    DownloadServerLogs downloadServerLogsTask = new DownloadServerLogs();
+    try {
+      return downloadServerLogsTask.execute().get();
```

Comment:
again, this is blocking, so there is no reason to put the code in another thread.

Reply:
will address in a seperate PR